src/ensembl/src/dataThunk.ts

```diff
@@ -0,0 +1,38 @@
+import { Dispatch } from 'redux';
+
+import config from 'config';
+import GoogleAnalyticsTracking from './services/analytics-service';
+import { buildAnalyticsObject, AnalyticsCategory } from './analyticsHelper';
+
+export const getDataThunk = (
+  endpoint: string,
+  asyncActions: any,
+  analyticsCategory: AnalyticsCategory,
+  id?: string
+) => {
+  return (dispatch: Dispatch) => {
+    dispatch(asyncActions.request(id));
+
+    const startTime = Date.now();
+
+    return fetch(`${config.apiHost}${endpoint}`)
+      .then(
+        (response) => {
+          const endTime = Date.now();
+          const timeDiff = endTime - startTime;
+
+          GoogleAnalyticsTracking.trackRequest(
+            buildAnalyticsObject(analyticsCategory)({
+              action: asyncActions.request.getType(),
+              label: 'AJAX Request',
+              value: timeDiff
+            })
+          );
+
+          return response.json();
+        },
+        (error) => dispatch(asyncActions.failure(error))
+      )
+      .then((json) => dispatch(asyncActions.success(json)));
+  };
+};
```

Comment:
I am very conflicted about this file.

On the one hand, part of what's going on here — like concatinating the host name with the pathname, doing the fetch itself, and then figuring out how to provide the response (i.e. that the json method has to be called) — seems to belong to a dedicated `apiService`. Especially since the request to GoogleAnalytics is then made not through our redux middleware (which means that this could also be a part of the service).

It also seems that we are putting data about individual response time in Google Analytics, while the proper place for it would be a monitoring service (which could also send alerts if responses start taking too long time).

Personally, I wouldn't even extract the thunk into a separate file. Especially considering that, if we extract the `apiService`, the code of the thunk itself becomes very small, and we might need all the flexibility it can offer, as in this example:

```
const someAsyncAction = () => async (dispatch, getState) => {
  dispatch(asyncAction.request());
  const someStateValie = someSelector(getState());
  try {
    const response = await apiService.fetch(endpoint, options);
    const modifiedResponse = modify(response);
    dispatch(asyncActions.success(modifiedResponse));
  } catch (error) {
    dispatch(asyncActions.failure(error))
  }
}
```

Reply:
The main reason why I extracted this code into a separate thunk file was that it was tightly coupled with the analytics. However, if we use a monitoring service for APIs rather than analytics (which I also think is much better), then the code can be easily put into an API service instead.